### PR TITLE
[wip] Expand destructive substitution

### PR DESCRIPTION
**Pull request by greedy**
Work in progress to address #96. I think it's producing the right sort of results but I'm not at all sure there aren't some corner cases that are incorrectly handled. I've made a test case but not any expectation files until I'm sure what the right result should be. Of course any comments about the implementation qualities are welcome as well.

I plan to squash this and remove the debugging before this would be merged.

### CODE REVIEW

**Comment by jonludlam** [comment]
This is looking really good thanks! I'll make some comments inline.

File: src/model/dump.ml [hunk 1]
```diff
@@ -0,0 +1,271 @@
+module Identifier =struct
```
Comment (jonludlam) [inline src/model/dump.ml]: This module is very similar to src/test/print/print.ml - if you've got anything extra can you merge it into there please.

File: src/xref/subst.ml [hunk 2]
```diff
@@ -371,3 +371,161 @@ end
...
+  method root x = x
+
+  method offset_identifier_signature x = x
+
+end
+
+let subst_type ~equal x y =
+  new subst_type ~equal x y
+
+class subst_module ~equal:_ (x:Path.Resolved.Module.t) (_y:Path.Module.t) : t = object (self)
```
Comment (jonludlam) [inline src/xref/subst.ml]: _y -> y

File: src/xref/subst.ml [hunk 3]
```diff
@@ -371,3 +371,161 @@ end
...
+        | _ ->
+          (match item with
+           | Lang.Signature.Value Lang.Value.{ id; type_=Lang.TypeExpr.Constr(tp, _); _ } ->
+             Printf.printf "Visiting value %s : _ %s\n" (Dump.Identifier.Value.dump id) (Dump.Path.Type.dump tp)
+           | _ -> ()
+          );
+          let item' = self#signature_item item in
+          item' :: items) s []
+
+  inherit Maps.paths as _super
```
Comment (jonludlam) [inline src/xref/subst.ml]: _super -> super

**Comment by greedy** [comment]
I'm not sure what to do to fix the test case I just added. My first thought was to rearrange the order of processing so that the substitution was done after expansion somehow. The problem here is that (for module substitution, at least) the substitution could target a module that is included and then maybe they'd need to be iterated to a fixed point to get completely correct results? Of course, it'd probably be okay to just approximate with subst-include-subst or something?

**Comment by jonludlam** [comment]
Hmm yes - very interesting! I'll have a poke around and see if I can come up with an explanation of what's going on.

**Comment by jonludlam** [comment]
I think I have an explanation of what's going on.

Right up at the top of `expand.ml` there's the `module_type` method:

```ocaml
method! module_type mty =
    let mty' = expand_module_type t mty in
      super#module_type mty'
```

The first bit of this essentially resolves anything that's not a signature already. In your case we've got two module types that are already signatures:

```ocaml
module type S = sig
  type 'a t
  val init : string t
end

module type X = sig
    include S
    val doit : 'a t -> unit
end
```

and one that isn't:

```ocaml
module type X2 = X with type 'a t := 'a list
```

The idea is that we turn `X` into `sig include S val doit : 'a t -> unit end` first, effectively by splicing in the definition of `X` into the `expansion` field of `X2` at this point and fixing up the identifiers by mapping `X -> X2`. We then recurse down this expansion in the call to `super#module_type`.

Unfortunately the way it's working now is that we do the call to `replace_type` in this first call to `expand_module_type`, which returns an expansion that has correctly had the type `t` erased. So at this point the expansion is now:

```ocaml
   Some
     (Module.Signature
       [Signature.Include
         {Include.parent =
           `ModuleType
             (`Root root,
              "X2");
          doc = [];
          decl =
           Module.ModuleType
            (ModuleType.Path
              (`Resolved
                 (`Identifier
                    (`ModuleType
                       (`Root root,
                        "S")))));
          expansion =
           {Include.resolved = false;
            content =
             [Signature.Value
               {Value.id =
                 `Value
                   (`ModuleType
                      (`Root root,
                       "X2"),
                    "init");
                doc = [];
                type_ =
                 TypeExpr.Constr
                  (`Resolved (`Identifier (`CoreType "list")),
                  [TypeExpr.Constr
                    (`Resolved (`Identifier (`CoreType "string")), [])])}]}};
        Signature.Value
         {Value.id =
           `Value
             (`ModuleType
                (`Root root,
                 "X2"),
              "doit");
          doc = [];
          type_ =
           TypeExpr.Arrow (None,
            TypeExpr.Constr
             (`Resolved (`Identifier (`CoreType "list")),
             [TypeExpr.Var "a"]),
            TypeExpr.Constr
             (`Resolved (`Identifier (`CoreType "unit")), []))}])}                   
```

Unfortunately in the `Include` the `expansion` field has `resolved=false`, so when we recurse through this expansion we hit:

```ocaml
let expand_include t incl =
  let open Include in
    if incl.expansion.resolved then incl
    else begin
      let root = Identifier.Signature.root incl.parent in
        match expand_include t root incl with
        | Expanded content' ->
            let content = remove_docs_from_signature content' in
            let expansion = {content;resolved=true} in
              { incl with expansion }
        | _ -> incl
    end
```

and because `resolved=false` we _reexpand_ the include, unfortunately skipping out the bit where we replace the type. So we end up with an expansion that has `X` respliced back in (though we do at least remember to rename it `X2`).

I'm not yet sure of the correct fix.

**Comment by lpw25** [comment]
I haven't had a chance to look at any of this yet, but as a hint: IIRC when an include expansion has `resolved = false` it has come from the `cmti` file directly not from the work of the actual expander. It is supposed to be replaced by the proper expansion done by the expander and should only actually be used for anything if resolving the path of the include fails for some reason.

**Comment by jonludlam** [comment]
I see, yes - the expansion of the `include` is loaded directly from the cmti file. We've already processed the expansion of `X` at this point but when we look up the path to `X` we get if from the original cmti rather than what we've already processed. So this could be a candidate for fixing with the improved expansion-ready components data structure?

**Comment by lpw25** [comment]
> So this could be a candidate for fixing with the improved expansion-ready components data structure?

I think so yes. IIRC the current expansion code does nothing for destructive substitution, whilst the current resolving code does something that only works in the common case. The hope is that the new components data structure would do the correct thing and so solve both of these issues.

**Comment by jonludlam** [comment]
This has been superseded by the work going on [here](https://github.com/jonludlam/odoc/tree/next)

**Comment by jonludlam** [comment]
Thanks for the work you put in on this @greedy - the new model should be much more understandable and easier to hack on, so I look forward to your future contributions :-) Also this particular issue is fixed!
